src/cargo/ops/tree/graph.rs

```diff
@@ -0,0 +1,562 @@
+use crate::core::dependency::DepKind;
+use crate::core::resolver::features::{FeaturesFor, RequestedFeatures, ResolvedFeatures};
+use crate::core::resolver::Resolve;
+use crate::core::{
+    FeatureMap, FeatureValue, InternedString, Package, PackageId, PackageIdSpec, Workspace,
+};
+use crate::util::CargoResult;
+use std::collections::{HashMap, HashSet};
+
+#[derive(Debug, Clone, Eq, PartialEq, Hash, Ord, PartialOrd)]
+pub enum Node {
+    Package {
+        package_id: PackageId,
+        /// Features that are enabled on this package.
+        features: Vec<InternedString>,
+        kind: CompileKind,
+    },
+    Feature {
+        /// Index of the package node this feature is for.
+        node_index: usize,
+        /// Name of the feature.
+        name: InternedString,
+    },
+}
+
+#[derive(Debug, Copy, Hash, Eq, Clone, PartialEq)]
+pub enum Edge {
+    Dep(DepKind),
+    Feature,
+}
+
+#[derive(Clone)]
+struct Edges(HashMap<Edge, Vec<usize>>);
+
+impl Edges {
+    fn new() -> Edges {
+        Edges(HashMap::new())
+    }
+
+    fn add_edge(&mut self, kind: Edge, index: usize) {
+        let indexes = self.0.entry(kind).or_default();
+        if !indexes.contains(&index) {
+            indexes.push(index)
+        }
+    }
+}
+
+/// A graph of dependencies.
+pub struct Graph<'a> {
+    nodes: Vec<Node>,
+    edges: Vec<Edges>,
+    /// Index maps a node to an index, for fast lookup.
+    index: HashMap<Node, usize>,
+    /// Map for looking up packages.
+    package_map: HashMap<PackageId, &'a Package>,
+    /// Set of indexes of feature nodes that were added via the command-line.
+    ///
+    /// For example `--features foo` will mark the "foo" node here.
+    cli_features: HashSet<usize>,
+    /// Map of dependency names, used for building internal feature map for
+    /// dep_name/feat_name syntax.
+    ///
+    /// Key is the index of a package node, value is a map of dep_name to a
+    /// set of `(pkg_node_index, is_optional)`.
+    dep_name_map: HashMap<usize, HashMap<InternedString, HashSet<(usize, bool)>>>,
+}
+
+impl<'a> Graph<'a> {
+    fn new(package_map: HashMap<PackageId, &'a Package>) -> Graph<'a> {
+        Graph {
+            nodes: Vec::new(),
+            edges: Vec::new(),
+            index: HashMap::new(),
+            package_map,
+            cli_features: HashSet::new(),
+            dep_name_map: HashMap::new(),
+        }
+    }
+
+    /// Adds a new node to the graph, returning its new index.
+    fn add_node(&mut self, node: Node) -> usize {
+        let from_index = self.nodes.len();
+        self.nodes.push(node);
+        self.edges.push(Edges::new());
+        self.index
+            .insert(self.nodes[from_index].clone(), from_index);
+        from_index
+    }
+
+    /// Returns a list of nodes the given node index points to for the given kind.
+    ///
+    /// Returns None if there are none.
+    pub fn connected_nodes(&self, from: usize, kind: &Edge) -> Option<Vec<usize>> {
```

Comment:
Take this with a grain of salt since I'm still reading, but could this return `Vec<usize>` instead of `Option<Vec<_>>` where empty means "no connections"?

Reply:
Yea, originally it was returning a reference and couldn't easily return an empty vec, but I later changed it (when I added sorting).